### ISSUE TITLE
WebAssembly aka WASM support?

### ISSUE THREAD

**Issue by JounQin** [hadolint/hadolint#873]
- [ ] This is a bug report
- [x] This is a feature request
- [ ] I searched existing issues before opening this one

### Expected behavior

As title

### Actual behavior

N/A

### Steps to reproduce the behavior

**Output of `hadolint --version` or
  `docker run --rm hadolint/hadolint hadolint --version` or
  `docker run --rm ghcr.io/hadolint/hadolint hadolint --version`:**

```bash
(paste your output here)
```

**Dockerfile** (if relevant)

```Dockerfile
(paste content of your Dockerfile)
```

**Additional environment details (OS, stack version, etc.)**

**Comment by ColdHeat** [comment]
I was curious about this awhile ago. Since the hadolint web site works in your browser without any extra requests you could in theory make something in wasm based off of that: https://github.com/hadolint/hadolint/tree/gh-pages.

**Comment by m-ildefons** [comment]
Hey,
Sorry for getting back to you so late. I think I now understand the initial question here.

The hadolint web version was created by Haskell-to-Javascript transpiler, but as far as I can tell the one used has been abandoned and there is no replacement compatible with modern GHC 9.x
There is however a [WASM backend for GHC](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9204).
I have no experience with WASM at this point, but you're more than welcome to contribute.

**Comment by m-ildefons** [comment]
I've looked a bit into compiling Hadolint with GHC with the WASM backend. The compiler as well as the entire Haskell ecosystem's support for WASM is still in its infancy at this point, however with a few hacks and fixes to various libraries (removing Template Haskell from ShellCheck amongst others), I was able to compile Hadolint. Unfortunately it couldn't execute successfully. I'm in contact with the upstream maintainers to work out where the problem lies and how to fix it:
https://gitlab.haskell.org/ghc/ghc/-/issues/23190

**Comment by anuraaga** [comment]
Just a light share, I was able to compile this to wasm with the following patch

https://github.com/wasilibs/go-hadolint/blob/main/buildtools%2Fwasm%2Fpatch.txt

The only incompatibility was `basement` via `cryptonite`. I found a lot of code there that didn't seem compatible with the wasm model of 64-bit numbers, 32-bit pointers, and I think it may be because of legacy coding patterns. I just removed the code that uses it which is one CI output formatter. If there are more modern libraries for computing digests, perhaps they would unlock making this library buildable without patches. Note, I don't know Haskell at all and am just trying to compile existing code to wasi so not sure if that makes any sense.

**Comment by anuraaga** [comment]
Sent #1090 to replace the old crypto library and could verify all the source builds fine for Wasm after just patching out the threading flags. Hopefully the project accepts it!
